From 11365c8e2caf1bdbabf4b3bf812d24a83099d41b Mon Sep 17 00:00:00 2001
From: Sean Paul <seanpaul@chromium.org>
Date: Tue, 9 Jan 2018 13:53:13 -0500
Subject: [PATCH 0534/1240] UPSTREAM: drm/i915: Only disable HDCP when it's
 active

Instead of always trying to disable HDCP. Only run hdcp_disable when the
state is not UNDESIRED. This will catch cases where it's enabled and
also cases where enable failed and the state is left in DESIRED mode.

Note that things won't blow up if disable is attempted while already
disabled, it's just bad form.

Reviewed-by: Daniel Vetter <daniel@ffwll.ch>
Signed-off-by: Sean Paul <seanpaul@chromium.org>
Link: https://patchwork.freedesktop.org/patch/msgid/20180109185330.16853-1-seanpaul@chromium.org
(cherry picked from commit 3ffdde8a3a8dd089578469eb0d79a30d5e81ec87)

Change-Id: I3ee42c62b50674a653705e592efa1913dcf7fefa
---
 drivers/gpu/drm/i915/intel_hdcp.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_hdcp.c b/drivers/gpu/drm/i915/intel_hdcp.c
index 3c164a2..827cab2 100644
--- a/drivers/gpu/drm/i915/intel_hdcp.c
+++ b/drivers/gpu/drm/i915/intel_hdcp.c
@@ -616,15 +616,17 @@ int intel_hdcp_enable(struct intel_connector *connector)
 
 int intel_hdcp_disable(struct intel_connector *connector)
 {
-	int ret;
+	int ret = 0;
 
 	if (!connector->hdcp_shim)
 		return -ENOENT;
 
 	mutex_lock(&connector->hdcp_mutex);
 
-	connector->hdcp_value = DRM_MODE_CONTENT_PROTECTION_UNDESIRED;
-	ret = _intel_hdcp_disable(connector);
+	if (connector->hdcp_value != DRM_MODE_CONTENT_PROTECTION_UNDESIRED) {
+		connector->hdcp_value = DRM_MODE_CONTENT_PROTECTION_UNDESIRED;
+		ret = _intel_hdcp_disable(connector);
+	}
 
 	mutex_unlock(&connector->hdcp_mutex);
 	cancel_delayed_work_sync(&connector->hdcp_check_work);
-- 
2.7.4

